proposals/stdlib/mutablelist-move.md

```diff
@@ -0,0 +1,145 @@
+
+* Items can be searched in a `MutableList` by having the matching items moved to the top of the `MutableList`, rather than creating a new `List` or some filtering abstraction
+
+* Observe this [demo application](https://github.com/thomasnield/rxkotlinfx-tornadofx-demo) to see several of these use cases in action.
+
+## Alternatives
+
+Developers could implement the extension functions themselves, as shown below. Moving items can be tricky and verbose to coordinate with local implementations.
+
+```kotlin
+/**
+ * Moves the given **T** item to the specified index
+ */
+fun <T> MutableList<T>.move(item: T, newIndex: Int)  {
+    val currentIndex = indexOf(item)
+    if (currentIndex < 0) return
+    removeAt(currentIndex)
+    add(newIndex, item)
+}
+
+/**
+ * Moves the given item at the `oldIndex` to the `newIndex`
+ */
+fun <T> MutableList<T>.move(oldIndex: Int, newIndex: Int)  {
+    val item = this[oldIndex]
+    removeAt(oldIndex)
+    add(newIndex, item)
+}
+
+/**
+ * Moves all items meeting a predicate to the given index
+ */
+fun <T> MutableList<T>.moveAll(newIndex: Int, predicate: (T) -> Boolean) {
+
+    var newIndexIncrement = newIndex
+
+    asSequence().toList().asSequence().withIndex()
+            .filter { v -> predicate.invoke(v.value) }
+            .forEach { move(it.index, newIndexIncrement++) }
+}
+
+/**
+ * Moves the given element at specified index up the **MutableList** by one increment
+ * unless it is at the top already which will result in no movement
+ */
+fun <T> MutableList<T>.moveUpAt(index: Int) {
+    if (index == 0) return
+    if (index < 0 || index >= size) throw Exception("Invalid index $index for MutableList of size $size")
+    val newIndex = index + 1
+    val item = this[index]
+    removeAt(index)
+    add(newIndex, item)
+}
+
+/**
+ * Moves the given element **T** up the **MutableList** by one increment
+ * unless it is at the bottom already which will result in no movement
+ */
+fun <T> MutableList<T>.moveDownAt(index: Int) {
+    if (index == size - 1) return
+    if (index < 0 || index >= size) throw Exception("Invalid index $index for MutableList of size $size")
+    val newIndex = index - 1
+    val item = this[index]
+    removeAt(index)
+    add(newIndex, item)
+}Provide options to solve them.
+
+/**
+ * Moves the given element **T** up the **MutableList** by an index increment
+ * unless it is at the top already which will result in no movement
+ */
+fun <T> MutableList<T>.moveUp(item: T) {
+    val currentIndex = indexOf(item)
+    val newIndex = (currentIndex - 1)
+    if (currentIndex <=0) return
+    remove(item)
+    add(newIndex, item)
+}
+
+/**
+ * Moves the given element **T** up the **MutableList** by an index increment
+ * unless it is at the bottom already which will result in no movement
+ */
+fun <T> MutableList<T>.moveDown(item: T) {
+    val currentIndex = indexOf(item)
+    val newIndex = (currentIndex + 1)
```

Comment:
So maybe a `false` needs to be returned to indicate a failure, or even throw an `Exception`.